File: typescript/creation-patterns/builder.js
var Product = /** @class */ (function () {
    function Product() {
        this.parts = [];
    }
    Product.prototype.listParts = function () {
        console.log("Product parts: ".concat(this.parts.join(', ')));
    };
    return Product;
}());
var ConcreteBuilder = /** @class */ (function () {
    function ConcreteBuilder() {
        this.reset();
    }
    ConcreteBuilder.prototype.reset = function () {
        this.product = new Product();
    };
    ConcreteBuilder.prototype.partA = function () {
        this.product.parts.push('A');
    };
    ConcreteBuilder.prototype.partB = function () {
        this.product.parts.push('B');
    };
    ConcreteBuilder.prototype.partC = function () {
        this.product.parts.push('C');
    };
    ConcreteBuilder.prototype.build = function () {
        var result = this.product;
        this.reset(); // Reset for next build
        return result;
    };
    return ConcreteBuilder;
}());
var Director = /** @class */ (function () {
    function Director() {
    }
    Director.prototype.setBuilder = function (builder) {
        this.builder = builder;
    };
    Director.prototype.buildMinimalViableProduct = function () {
        this.builder.partA();
    };
    Director.prototype.buildFullFeaturedProduct = function () {
        this.builder.partA();
        this.builder.partB();
        this.builder.partC();
    };
    return Director;
}());
function clientCode(director) {
    var builder = new ConcreteBuilder();
    director.setBuilder(builder);
    director.buildMinimalViableProduct();
    builder.build().listParts();
    director.buildFullFeaturedProduct();
    builder.build().listParts();
}
var director = new Director();
clientCode(director);

File: rust/src/bin/creation-patterns/builder.rs
trait Bulder {
    fn part_a(&mut self);
    fn part_b(&mut self);
    fn part_c(&mut self);
}

#[derive(Clone)]
struct Product {
    parts: Vec<String>,
}

impl Product {
    fn list_parts(&self) {
        println!("Product parts: {:?}", self.parts.join(", "));
    }
}

struct ConcreteBuilder {
    product: Product,
}

impl ConcreteBuilder {
    fn new() -> Self {
        Self {
            product: Product { parts: Vec::new() },
        }
    }

    fn reset(&mut self) {
        self.product = Product { parts: Vec::new() };
    }

    fn build(&mut self) -> Product {
        let result = self.product.clone();

        self.reset();

        result
    }
}

impl Bulder for ConcreteBuilder {
    fn part_a(&mut self) {
        self.product.parts.push("A".to_string());
    }

    fn part_b(&mut self) {
        self.product.parts.push("B".to_string());
    }

    fn part_c(&mut self) {
        self.product.parts.push("C".to_string());
    }
}

struct Director;

impl Director {
    fn build_minimal_viable_product(builder: &mut impl Bulder) {
        builder.part_a();
    }

    fn build_full_featured_product(builder: &mut impl Bulder) {
        builder.part_a();
        builder.part_b();
        builder.part_c();
    }
}

fn main() {
    let mut builder = ConcreteBuilder::new();

    Director::build_minimal_viable_product(&mut builder);
    builder.build().list_parts();

    Director::build_full_featured_product(&mut builder);
    builder.build().list_parts();
}
